### PR TITLE
Pass logger into all middlewares to log warnings on 5xx

### PR DESCRIPTION
**Pull request by metalmatze**
Additionally, I've added request IDs to the request context and all middlewares / handlers are going to log with that request ID to correlate log lines for individual requests.

I went through all `http.Error` calls and made sure to add a log line with it. Everything that's a 4xx response is going to `level.Debug` and 5xx are going to `level.Warn`. 

/cc @squat @kakkoyun @brancz

### CODE REVIEW

File: pkg/server/forward.go [hunk 1]
```diff
@@ -134,7 +149,9 @@ func ForwardHandler(logger log.Logger, forwardURL *url.URL) http.HandlerFunc {
 
 		if resp.StatusCode/100 != 2 {
 			// surfacing upstreams error to our users too
-			http.Error(w, fmt.Errorf("response status code is %s", resp.Status).Error(), resp.StatusCode)
+			msg := fmt.Sprintf("response status code is %s", resp.Status)
+			level.Warn(logger).Log("msg", msg)
```
[the remark this answers is not in the document]
Comment (metalmatze) [reply pkg/server/forward.go:153]: Good point. I guess for now it's fine. If it becomes too spammy, we might want to add another if in there :)

**Comment by metalmatze** [comment]
Seems like it was missing a `go mod vendor`. Thanks for the pointer @kakkoyun.

**Comment by brancz** [comment]
This only touches server side telemetry code merging so we can debug our production system :)

**Pull request merged**